### PR TITLE
SelectionList fixes

### PR DESCRIPTION
**Pull request by alisey**
* Items can be filtered
* dataMapper support
* The list scrolls to focused item
* Keyboard navigations works with multiple identical values in the list
* Clicking on a disabled item doesn't focus the list

### CODE REVIEW

**Comment by JoMarton** [comment]
@alisey  In general - I would have expected to see at least one test added for each of the above features mentioned in the description (for bugs maybe just amending an existing test). Can additional tests be added?

File: src/components/selection-list/selection-list-model.tsx [hunk 1]
```diff
@@ -1,166 +1,266 @@
...
+        if (oldIndex !== newIndex) {
+            (oldIndex > -1) && (this.items[oldIndex].selected = false);
+            (newIndex > -1) && (this.items[newIndex].selected = true);
+            this.selectedIndex = newIndex;
+        }
     }
 
-    private focusIndex(index: number) {
-        this.focusValue(this.selectableValues[clamp(index, 0, this.selectableValues.length - 1)]);
+    private findSelectable(startIndex: number, step: number) {
```
Comment (JoMarton) [inline src/components/selection-list/selection-list-model.tsx]: return type?

File: src/components/selection-list/selection-list-model.tsx [hunk 2]
```diff
@@ -1,166 +1,266 @@
...
+        return this.selectedIndex > -1;
+    }
+
+    @action public focusSelected() {
+        if (this.selectedIndex > -1 && this.items[this.selectedIndex].selectable) {
+            this.focusIndex(this.selectedIndex);
+        }
+    }
+
+    public typeAhead(character: string) {
```
Comment (JoMarton) [inline src/components/selection-list/selection-list-model.tsx]: is there a reason to have this method available?

File: src/components/selection-list/selection-list-model.tsx [hunk 3]
```diff
@@ -1,166 +1,266 @@
...
-            this.selectableValues.push(item.value);
+    @action public focusIndex(newIndex: number): boolean {
+        const oldIndex = this.focusedIndex;
+        if (oldIndex !== newIndex) {
+            (oldIndex > -1) && (this.items[oldIndex].focused = false);
+            (newIndex > -1) && (this.items[newIndex].focused = true);
+            this.focusedIndex = newIndex;
+            return true;
         }
+        return false;
```
Comment (JoMarton) [inline src/components/selection-list/selection-list-model.tsx]: tested?

File: src/components/selection-list/selection-list-model.tsx [hunk 4]
```diff
@@ -1,166 +1,266 @@
...
-        return this.items.find(item => item.value === value);
+    constructor(public item: React.ReactChild) {}
+
+    public get isOption() {
+        return typeof this.item === 'object' && this.item.props.value !== undefined;
     }
 
-    public getSelectedValue(): SelectionListItemValue | undefined {
-        return this.selectedValue;
+    public get selectable() {
```
Comment (JoMarton) [inline src/components/selection-list/selection-list-model.tsx]: return type?

File: src/components/selection-list/selection-list-model.tsx [hunk 5]
```diff
@@ -1,166 +1,266 @@
...
 
-    public focusFirst() {
-        this.focusIndex(0);
+    public get value() {
+        return this.itemResolved.value;
     }
 
-    public focusLast() {
-        this.focusIndex(Infinity);
+    public get label() {
```
Comment (JoMarton) [inline src/components/selection-list/selection-list-model.tsx]: return type

File: src/components/selection-list/selection-list-model.tsx [hunk 6]
```diff
@@ -1,166 +1,266 @@
...
+        return this.itemResolved.value !== undefined;
+    }
+
+    public get selectable() {
+        return this.isOption && !this.disabled;
     }
 
-    public focusFirst() {
-        this.focusIndex(0);
+    public get value() {
```
Comment (JoMarton) [inline src/components/selection-list/selection-list-model.tsx]: can return type for value be specified here?

File: src/components/selection-list/selection-list-model.tsx [hunk 7]
```diff
@@ -1,166 +1,266 @@
...
-        React.Children.forEach(children, element => {
-            if (typeof element === 'object') {
-                this.addItem(element, element);
-            }
-        });
+    public get isOption() {
+        return this.itemResolved.value !== undefined;
+    }
+
+    public get selectable() {
```
Comment (JoMarton) [inline src/components/selection-list/selection-list-model.tsx]: return type

File: src/components/selection-list/selection-list-model.tsx [hunk 8]
```diff
@@ -1,166 +1,266 @@
...
+        this.itemResolved = resolveItem(item);
     }
 
-    public addChildren(children: React.ReactNode) {
-        React.Children.forEach(children, element => {
-            if (typeof element === 'object') {
-                this.addItem(element, element);
-            }
-        });
+    public get isOption() {
```
Comment (JoMarton) [inline src/components/selection-list/selection-list-model.tsx]: return type

File: src/components/selection-list/selection-list-model.tsx [hunk 9]
```diff
@@ -1,166 +1,266 @@
...
-    public focusLast() {
-        this.focusIndex(Infinity);
+    public get label() {
+        return this.itemResolved.label || '';
     }
 
-    public focusPrevious() {
-        const currentIndex = this.getFocusedIndex();
-        this.focusIndex(currentIndex === -1 ? Infinity : currentIndex - 1);
+    public get disabled() {
```
Comment (JoMarton) [inline src/components/selection-list/selection-list-model.tsx]: return type

File: src/components/selection-list/selection-list-model.tsx [hunk 10]
```diff
@@ -1,166 +1,266 @@
...
+
+class SelectionListItemBasedOnChild implements SelectionListItem {
+    @observable public selected: boolean = false;
+    @observable public focused: boolean = false;
 
-    public findItemByValue(value: SelectionListItemValue): SelectionListItem | undefined {
-        return this.items.find(item => item.value === value);
+    constructor(public item: React.ReactChild) {}
+
+    public get isOption() {
```
Comment (JoMarton) [inline src/components/selection-list/selection-list-model.tsx]: return type

File: src/components/selection-list/selection-list-model.tsx [hunk 11]
```diff
@@ -1,166 +1,266 @@
...
 
-    public getSelectedValue(): SelectionListItemValue | undefined {
-        return this.selectedValue;
+    public get selectable() {
+        return  this.isOption && !this.disabled;
     }
 
-    public getFocusedValue(): SelectionListItemValue | undefined {
-        return this.focusedValue;
+    public get value() {
```
Comment (JoMarton) [inline src/components/selection-list/selection-list-model.tsx]: return type

File: src/components/selection-list/selection-list-model.tsx [hunk 12]
```diff
@@ -1,166 +1,266 @@
...
+        return typeof this.item === 'object' ? this.item.props.value : undefined;
     }
 
-    public selectValue(value: SelectionListItemValue | undefined) {
-        this.selectedValue = value;
-        for (const item of this.items) {
-            item.selected = (value !== undefined && item.value === value);
+    public get label() {
```
Comment (JoMarton) [inline src/components/selection-list/selection-list-model.tsx]: return type

File: src/components/selection-list/selection-list-model.tsx [hunk 13]
```diff
@@ -1,166 +1,266 @@
...
+                return this.item.props.label;
+            }
+            if (typeof this.item.props.children === 'string') {
+                return this.item.props.children;
+            }
         }
+        return '';
+    }
+
+    public get disabled() {
```
Comment (JoMarton) [inline src/components/selection-list/selection-list-model.tsx]: return type

File: src/components/selection-list/option.tsx [hunk 14]
```diff
@@ -7,14 +7,19 @@ export interface OptionProps {
...
+    // The value is not used by the option component, it only exists to expose the option's value to the list model.
     value?: SelectionListItemValue;
 }
 
 export const SelectionListOption: React.SFC<OptionProps> = stylable(style)(
     props => (
         <div
-            data-value={props.disabled ? undefined : props.value}
+            id={props.id}
+            role="option"
```
Comment (JoMarton) [inline src/components/selection-list/option.tsx]: verify in test that we are accessible

File: src/components/selection-list/selection-list-model.tsx [hunk 15]
```diff
@@ -1,166 +1,266 @@
...
-    public focusFirst() {
-        this.focusIndex(0);
+    public get value() {
+        return this.itemResolved.value;
     }
 
-    public focusLast() {
-        this.focusIndex(Infinity);
+    public get label() {
+        return this.itemResolved.label || '';
```
[the remark this answers is not in the document]
Comment (alisey) [reply src/components/selection-list/selection-list-model.tsx]: @JoMarton What test do you think needs to be added?

We have this:
https://github.com/wix/stylable-components/blob/master/test/components/selection-list.spec.tsx#L55

And this:
https://github.com/wix/stylable-components/blob/master/test/components/selection-list.spec.tsx#L65

[the remark this answers is not in the document]
Comment (alisey) [reply src/components/selection-list/selection-list-model.tsx]: Oh, I see now why the tests don't fail, this property is only used for typeahead, which is not implemented.

File: src/components/selection-list/selection-list-model.tsx (hunk 2 above)
Comment (alisey) [reply src/components/selection-list/selection-list-model.tsx]: Just as a reminder that this is where it should go.

**Comment by alisey** [comment]
Added:

* Return types
* hasSelection test
* accessibility tests
* onMouseDown test
* scrolling tests

**Pull request merged**